libs/entitlement/src/main/java/org/elasticsearch/entitlement/runtime/policy/PolicyManager.java

```diff
@@ -56,27 +84,120 @@ public PolicyManager(Policy defaultPolicy, Map<String, Policy> pluginPolicies, F
         throw new NotEntitledException(
-            Strings.format("Missing entitlement [%s] for caller [%s] in module [%s]", type, callerClass, requestingModule.getName())
+            Strings.format(
+                "Caller [%s] in module [%s] does not have [%s]",
+                callerClass,
+                requestingModule.getName(),
+                entitlementClass.getSimpleName()
+            )
         );
     }
 
+    ModuleEntitlements getEntitlementsOrThrow(Class<?> callerClass, Module requestingModule) {
+        ModuleEntitlements cachedEntitlement = moduleEntitlementsMap.get(requestingModule);
+        if (cachedEntitlement != null) {
+            if (cachedEntitlement == ModuleEntitlements.NONE) {
+                throw new NotEntitledException(buildModuleNoPolicyMessage(callerClass, requestingModule) + "[CACHED]");
+            }
+            return cachedEntitlement;
+        }
+
+        if (isServerModule(requestingModule)) {
+            var scopeName = requestingModule.getName();
+            return getModuleEntitlementsOrThrow(callerClass, requestingModule, serverPolicy, scopeName, "server");
+        }
+
+        // plugins
+        var pluginName = pluginResolver.apply(callerClass);
+        if (pluginName != null) {
+            var pluginPolicy = pluginPolicies.get(pluginName);
+            if (pluginPolicy != null) {
+                final String scopeName;
+                if (requestingModule.isNamed() == false) {
+                    scopeName = ALL_UNNAMED;
+                } else {
+                    scopeName = requestingModule.getName();
+                }
+                return getModuleEntitlementsOrThrow(callerClass, requestingModule, pluginPolicy, scopeName, pluginName);
+            }
+        }
+
+        moduleEntitlementsMap.put(requestingModule, ModuleEntitlements.NONE);
+        throw new NotEntitledException(buildModuleNoPolicyMessage(callerClass, requestingModule));
+    }
+
+    private static String buildModuleNoPolicyMessage(Class<?> callerClass, Module requestingModule) {
+        return Strings.format("Caller [%s] in module [%s] does not have any policy", callerClass, requestingModule.getName());
+    }
+
+    private ModuleEntitlements getModuleEntitlementsOrThrow(
+        Class<?> callerClass,
+        Module module,
+        Policy policy,
+        String moduleName,
+        String codebase
+    ) {
+        var entitlements = lookupEntitlementsForModule(policy, moduleName);
+        if (entitlements == null) {
+            // Module without entitlements - remember we don't have any
+            logger.warn(
+                "Class {} from module {} in {} does not have a policy; caching as unresolvable",
+                callerClass.getName(),
+                moduleName,
+                codebase
+            );
+            moduleEntitlementsMap.put(module, ModuleEntitlements.NONE);
+            throw new NotEntitledException(buildModuleNoPolicyMessage(callerClass, module));
```

Comment:
Nit: Since we're throwing here, the `warn` seems a bit redundant. This is probably going to kill the server anyway, or at the very least, it will get logged someplace.

If you do want to `warn` at this location, consider creating the exception object first and including it in the warning message so we get a stack trace.